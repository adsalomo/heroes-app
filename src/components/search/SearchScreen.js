import React, { useMemo } from 'react';
import queryString from 'query-string';

import { useForm } from '../../hooks/useForm';
import { HeroCard } from '../heroes/HeroCard';
import { useLocation } from 'react-router-dom';
import { getHeroesByName } from '../../selectors/getHeroesByName';

export const SearchScreen = ({ history }) => {

    const location = useLocation();

    const { q = '' } = queryString.parse(location.search);

    const [{ search }, handleInputChange] = useForm({
        search: q
    });

    const heroesFiltered = useMemo(() => getHeroesByName(q), [q]);

    const handleSearch = (e) => {
        e.preventDefault();
        history.push(`?q=${search}`);
    }

    return (
        <div>
            <h1>Search</h1>
            <hr />

            <div className="row">
                <div className="col-5">
                    <h4>Search form</h4>
                    <hr />
                    <form onSubmit={handleSearch}>
                        <input
                            onChange={handleInputChange}
                            name="search"
                            value={search ? search : ''}
                            className="form-control"
                            placeholder="Search"
                            autoComplete="off"
                            type="text" />
                        <button
                            type="submit"
                            className="btn btn-primary mt-3"
                        >
                            Search
                        </button>
                    </form>
                </div>

                <div className="col-7">
                    <h4>Resultado</h4>
                    <hr />
                    {
                        q === ''
                        &&
                        <div className="alert alert-info">
                            Por favor busque un heroe
                        </div>
                    }

                    {
                        q !== '' && heroesFiltered.length === 0
                        &&
                        <div className="alert alert-danger">
                            No existe informaci√≥n por {q}
                        </div>
                    }

                    {
                        heroesFiltered.map(hero => <HeroCard key={hero.id} {...hero} />)
                    }
                </div>
            </div>
        </div>
    )
}
