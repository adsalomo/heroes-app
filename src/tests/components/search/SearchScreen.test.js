import { mount } from "enzyme";
import { MemoryRouter, Route } from "react-router-dom";
import { SearchScreen } from "../../../components/search/SearchScreen";

describe('Prueba SearchScreen', () => {

    test('Debe mostrarse correctamente', () => {
        const wrapper = mount(
            <MemoryRouter
                initialEntries={['/search']}
            >
                <Route path="/search" component={SearchScreen} />
            </MemoryRouter>
        );

        expect(wrapper).toMatchSnapshot();
    });

    test('Debe mostrarse a batman y el input con el valor de queryString', () => {
        const wrapper = mount(
            <MemoryRouter
                initialEntries={['/search?q=batman']}
            >
                <Route path="/search" component={SearchScreen} />
            </MemoryRouter>
        );

        expect(wrapper.find('input').prop('value')).toBe('batman');

    });

    test('Debe mostrar un error sino encuentra un Hero', () => {
        const wrapper = mount(
            <MemoryRouter
                initialEntries={['/search?q=batman123']}
            >
                <Route path="/search" component={SearchScreen} />
            </MemoryRouter>
        );

        expect(wrapper.find('.alert-danger').exists()).toBe(true);

    });

    test('Debe de llamar el push del history', () => {
        const history = {
            push: jest.fn(),
        };

        const wrapper = mount(
            <MemoryRouter
                initialEntries={['/search?q=batman']}
            >
                <Route
                    path="/search"
                    component={() => <SearchScreen history={history} />}
                />
            </MemoryRouter>
        );

        wrapper.find('input').simulate('change', {
            target: {
                name: 'search',
                value: 'batman',
            }
        });

        wrapper.find('form').prop('onSubmit')({ preventDefault() { } });

        expect(history.push).toHaveBeenCalledWith(`?q=batman`);
    });

});
