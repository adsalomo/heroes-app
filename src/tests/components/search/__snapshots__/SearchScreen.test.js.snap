// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prueba SearchScreen Debe mostrarse correctamente 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="search"
          onChange={[Function]}
          placeholder="Search"
          type="text"
          value=""
        />
        <button
          className="btn btn-primary mt-3"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Resultado
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Por favor busque un heroe
      </div>
    </div>
  </div>
</div>
`;
